readme.md
# League\Flysystem\WebDAV [BETA]

[![Author](http://img.shields.io/badge/author-@frankdejonge-blue.svg?style=flat-square)](https://twitter.com/frankdejonge)
[![Build Status](https://img.shields.io/travis/thephpleague/flysystem-webdav/master.svg?style=flat-square)](https://travis-ci.org/thephpleague/flysystem-webdav)
[![Coverage Status](https://img.shields.io/scrutinizer/coverage/g/thephpleague/flysystem-webdav.svg?style=flat-square)](https://scrutinizer-ci.com/g/thephpleague/flysystem-webdav)
[![Quality Score](https://img.shields.io/scrutinizer/g/thephpleague/flysystem-webdav.svg?style=flat-square)](https://scrutinizer-ci.com/g/thephpleague/flysystem-webdav)
[![Software License](https://img.shields.io/badge/license-MIT-brightgreen.svg?style=flat-square)](LICENSE)
<!--
[![Packagist Version](https://img.shields.io/packagist/v/league/flysystem-webdav.svg?style=flat-square)](https://packagist.org/packages/league/flysystem-webdav)
[![Total Downloads](https://img.shields.io/packagist/dt/league/flysystem-webdav.svg?style=flat-square)](https://packagist.org/packages/league/flysystem-webdav)
-->

This is a Flysystem adapter for the WebDAV.

## Installation

```bash
composer require league/flysystem-webdav
```

## Bootstrap

``` php
<?php
use Sabre\DAV\Client;
use League\Flysystem\Filesystem;
use League\Flysystem\WebDAV\WebDAVAdapter;

$client = new Client($settings);
$adapter = new WebDAVAdapter($client);
$flysystem = new Filesystem($adapter);
```


changelog.md
# Changelog

## 1.0.9 - xxx

* Made detection of directories compliant with rfc4918 (#53).

## 1.0.8 - 2019-07-08

* Made some properties private to allow them to be overwritten so other types of WebDAV servers can be supported (#51).

## 1.0.7 - 2019-06-23

* Fixed prefix handling from listContent

## 1.0.6 - 2018-12-14

### Fixed

* Create directories with a trailing "/".

## 1.0.5 - 2016-12-14

### Fixed

* Copy now also works with native webdav capabilities.

## 1.0.4 - 2016-08-11

### Fixed

* The getMetadata function now also catches http exceptions.

## 1.0.3 - 2016-07-11

### Improved

* The `sabre/dav` version is upped to ~3.1.

## 1.0.2 - 2016-03-17

### Improved

* Listing contents now prevents 301 redirects by adding a trailing slash.

## 1.0.1 - 2015-05-18

### Fixed

* Corrected namespace for missing object exception catching.
* Corrected last-modified return type handling (int -> [int])


## 1.0.0

Initial release


tests/WebDAVTests.php
<?php

use League\Flysystem\Config;
use League\Flysystem\Filesystem;
use League\Flysystem\WebDAV\WebDAVAdapter;

class WebDAVTests extends PHPUnit_Framework_TestCase
{
    protected function getClient()
    {
        return Mockery::mock('Sabre\DAV\Client');
    }

    public function testHas()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('propFind')->once()->andReturn([
            '{DAV:}getcontentlength' => 20,
        ]);
        $adapter = new Filesystem(new WebDAVAdapter($mock));
        $this->assertTrue($adapter->has('something'));
    }

    /**
     * @dataProvider provideExceptionsForHasFail
     */
    public function testHasFail($exceptionClass)
    {
        $mock = $this->getClient();
        $mock->shouldReceive('propFind')->once()->andThrow($exceptionClass);
        $adapter = new WebDAVAdapter($mock);
        $this->assertFalse($adapter->has('something'));
    }

    public function provideExceptionsForHasFail()
    {
        return [
            [Mockery::mock('Sabre\DAV\Exception\NotFound')],
            [Mockery::mock('Sabre\HTTP\ClientHttpException')],
        ];
    }

    public function testWrite()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->once()->andReturn([
            'statusCode' => 200,
        ]);
        $adapter = new WebDAVAdapter($mock);
        $this->assertInternalType('array', $adapter->write('something', 'something', new Config()));
    }

    public function testWriteFail()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->with('PUT', 'something', 'something')->once()->andReturn([
            'statusCode' => 500,
        ]);
        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->write('something', 'something', new Config());
        $this->assertFalse($result);
    }

    public function testWriteStream()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->once()->andReturn([
            'statusCode' => 200,
        ]);
        $adapter = new WebDAVAdapter($mock);

        $tmp = $this->getLargeTmpStream();

        $this->assertInternalType('array', $adapter->writeStream('something', $tmp, new Config()));

        if (is_resource($tmp)) {
            fclose($tmp);
        }
    }

    protected function getLargeTmpStream()
    {
        $size = intval($this->getMemoryLimit() * 1.5);
        $tmp = tmpfile();
        fseek($tmp, $size);
        fprintf($tmp, 'a');
        fflush($tmp);

        return $tmp;
    }

    protected function getMemoryLimit()
    {
        $unit_factor = [
            ''  => 0,
            'K' => 1,
            'M' => 2,
            'G' => 3,
        ];

        if (!preg_match("/^(\d+)([KMG]?)$/i", ini_get('memory_limit'), $match)) {
            throw new Exception('invalid memory_limit?');
        }

        $limit = $match[1] * pow(1024, $unit_factor[strtoupper($match[2])]);

        return $limit;
    }

    public function testUpdate()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->once();
        $adapter = new WebDAVAdapter($mock);
        $this->assertInternalType('array', $adapter->update('something', 'something', new Config()));
    }

    /**
     * @expectedException LogicException
     */
    public function testWriteVisibility()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->once()->andReturn([
            'statusCode' => 200,
        ]);
        $adapter = new WebDAVAdapter($mock);
        $this->assertInternalType('array', $adapter->write('something', 'something', new Config([
            'visibility' => 'private',
        ])));
    }

    public function testReadStream()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->andReturn([
            'statusCode' => 200,
            'body' => 'contents',
            'headers' => [
                'last-modified' => date('Y-m-d H:i:s'),
            ],
        ]);
        $adapter = new WebDAVAdapter($mock, 'bucketname', 'prefix');
        $result = $adapter->readStream('file.txt');
        $this->assertInternalType('resource', $result['stream']);
    }

    public function testRename()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->once()->andReturn([
            'statusCode' => 200,
        ]);
        $adapter = new WebDAVAdapter($mock, 'bucketname');
        $result = $adapter->rename('old', 'new');
        $this->assertTrue($result);
    }

    public function testRenameFail()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->once()->andReturn([
            'statusCode' => 404,
        ]);
        $adapter = new WebDAVAdapter($mock, 'bucketname');
        $result = $adapter->rename('old', 'new');
        $this->assertFalse($result);
    }

    public function testRenameFailException()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->once()->andThrow('Sabre\DAV\Exception\NotFound');
        $adapter = new WebDAVAdapter($mock, 'bucketname');
        $result = $adapter->rename('old', 'new');
        $this->assertFalse($result);
    }

    public function testDeleteDir()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->with('DELETE', 'some/dirname')->once()->andReturn(['statusCode' => 200]);
        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->deleteDir('some/dirname');
        $this->assertTrue($result);
    }

    public function testDeleteDirFailNotFound()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->with('DELETE', 'some/dirname')->once()->andThrow('Sabre\DAV\Exception\NotFound');
        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->deleteDir('some/dirname');
        $this->assertFalse($result);
    }

    public function testDeleteDirFailNot200Status()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->with('DELETE', 'some/dirname')->once()->andReturn(['statusCode' => 403]);
        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->deleteDir('some/dirname');
        $this->assertFalse($result);
    }

    public function testListContents()
    {
        $mock = $this->getClient();
        $first = [
            [],
            'filename' => [
                '{DAV:}getcontentlength' => "20",
                '{DAV:}iscollection' => "0",
            ],
            'dirname' => [
                '{DAV:}getcontentlength' => "0",
                '{DAV:}iscollection' => "1",
            ],
        ];

        $second = [
            [],
            'deeper_filename.ext' => [
                '{DAV:}getcontentlength' => "20",
                '{DAV:}iscollection' => "0",
            ],
        ];
        $mock->shouldReceive('propFind')->twice()->andReturn($first, $second);
        $adapter = new WebDAVAdapter($mock, 'bucketname');
        $listing = $adapter->listContents('', true);
        $this->assertInternalType('array', $listing);
    }

    public function testListContentsWithPlusInName()
    {
        $mock = $this->getClient();
        $first = [
            [],
            'bucketname/dirname+something' => [
                '{DAV:}getcontentlength' => "0",
                '{DAV:}iscollection' => "1",
            ],
        ];

        $mock->shouldReceive('propFind')->once()->andReturn($first);
        $adapter = new WebDAVAdapter($mock, 'bucketname');
        $listing = $adapter->listContents('', false);
        $this->assertInternalType('array', $listing);
        $this->assertCount(1, $listing);
        $this->assertEquals('dirname+something', $listing[0]['path']);
    }

    public function testListContentsWithUrlEncodedSpaceInName()
    {
        $mock = $this->getClient();
        $first = [
            [],
            '/My%20Library/New%20Record%201.mp3' => [
                '{DAV:}displayname' => "New Record 1.mp3",
                '{DAV:}getcontentlength' => "8223370",
            ],
        ];

        $mock->shouldReceive('propFind')->once()->andReturn($first);
        $adapter = new WebDAVAdapter($mock, '/My Library');
        $listing = $adapter->listContents('', false);
        $this->assertInternalType('array', $listing);
        $this->assertCount(1, $listing);
        $this->assertEquals('New Record 1.mp3', $listing[0]['path']);
        $this->assertEquals('file', $listing[0]['type']);
        $this->assertEquals('8223370', $listing[0]['size']);
    }

    public function methodProvider()
    {
        return [
            ['getMetadata'],
            ['getTimestamp'],
            ['getMimetype'],
            ['getSize'],
        ];
    }

    /**
     * @dataProvider  methodProvider
     */
    public function testMetaMethods($method)
    {
        $mock = $this->getClient();
        $mock->shouldReceive('propFind')->once()->andReturn([
            '{DAV:}displayname' => 'object.ext',
            '{DAV:}getcontentlength' => 30,
            '{DAV:}getcontenttype' => 'plain/text',
            '{DAV:}getlastmodified' => date('Y-m-d H:i:s'),
        ]);
        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->{$method}('object.ext');
        $this->assertInternalType('array', $result);
    }

    public function testCreateDir()
    {
        /** @var Sabre\DAV\Client|Mockery\Mock $mock */
        $mock = $this->getClient();

        $mock->shouldReceive('propFind')
            ->once()
            ->andThrow(new \Sabre\DAV\Exception('Not found'));

        $mock->shouldReceive('request')
            ->once()
            ->with('MKCOL', 'dirname/')
            ->andReturn([
                'statusCode' => 201,
            ]);

        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->createDir('dirname', new Config());
        $this->assertInternalType('array', $result);
    }

    public function testCreateDirRecursive()
    {
        /** @var Sabre\DAV\Client|Mockery\Mock $mock */
        $mock = $this->getClient();

        $mock->shouldReceive('propFind')
            ->times(2)
            ->andThrow(new \Sabre\DAV\Exception('Not found'));

        $mock->shouldReceive('request')
            ->once()
            ->with('MKCOL', 'dirname/')
            ->andReturn([
                'statusCode' => 201,
            ]);

        $mock->shouldReceive('request')
            ->once()
            ->with('MKCOL', 'dirname/subdirname/')
            ->andReturn([
                'statusCode' => 201,
            ]);

        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->createDir('dirname/subdirname', new Config());
        $this->assertInternalType('array', $result);
    }

    public function testCreateDirIfExists()
    {
        /** @var Sabre\DAV\Client|Mockery\Mock $mock */
        $mock = $this->getClient();

        $mock->shouldReceive('propFind')
            ->once()
            ->andReturn([
                '{DAV:}displayname' => 'dirname',
                '{DAV:}getcontentlength' => 30,
                '{DAV:}getcontenttype' => 'dir',
                '{DAV:}getlastmodified' => date('Y-m-d H:i:s'),
            ]);

        $mock->shouldReceive('request')
            ->never();

        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->createDir('dirname', new Config());
        $this->assertInternalType('array', $result);
    }

    public function testCreateDirFail()
    {
        /** @var Sabre\DAV\Client|Mockery\Mock $mock */
        $mock = $this->getClient();

        $mock->shouldReceive('propFind')
            ->once()
            ->andThrow(new \Sabre\DAV\Exception('Not found'));

        $mock->shouldReceive('request')
            ->once()
            ->with('MKCOL', 'dirname/')
            ->andReturn([
                'statusCode' => 500,
            ]);

        $adapter = new WebDAVAdapter($mock);
        $result = $adapter->createDir('dirname', new Config());
        $this->assertFalse($result);
    }

    public function testRead()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->andReturn([
            'statusCode' => 200,
            'body' => 'contents',
            'headers' => [
                'last-modified' => [date('Y-m-d H:i:s')],
            ],
        ]);
        $adapter = new WebDAVAdapter($mock, 'bucketname', 'prefix');
        $result = $adapter->read('file.txt');
        $this->assertInternalType('array', $result);
    }

    public function testReadFail()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->andReturn([
            'statusCode' => 404,
            'body' => 'contents',
            'headers' => [
                'last-modified' => [date('Y-m-d H:i:s')],
            ],
        ]);
        $adapter = new WebDAVAdapter($mock, 'bucketname', 'prefix');
        $result = $adapter->read('file.txt');
        $this->assertFalse($result);
    }

    public function testReadStreamFail()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->andReturn([
            'statusCode' => 404,
            'body' => 'contents',
            'headers' => [
                'last-modified' => [date('Y-m-d H:i:s')],
            ],
        ]);
        $adapter = new WebDAVAdapter($mock, 'bucketname', 'prefix');
        $result = $adapter->readStream('file.txt');
        $this->assertFalse($result);
    }

    public function testReadException()
    {
        $mock = $this->getClient();
        $mock->shouldReceive('request')->andThrow('Sabre\DAV\Exception\NotFound');
        $adapter = new WebDAVAdapter($mock, 'bucketname', 'prefix');
        $result = $adapter->read('file.txt');
        $this->assertFalse($result);
    }

    public function testNativeCopy()
    {
        /** @var Sabre\DAV\Client|Mockery\Mock $clientMock */
        $clientMock = $this->getClient();

        $clientMock->shouldReceive('getAbsoluteUrl')->andReturn('http://webdav.local/prefix/newFile.txt');

        $clientMock->shouldReceive('request')->andReturn([
            'statusCode' => 201
        ]);

        $adapter = new WebDAVAdapter($clientMock, 'prefix', false);
        $result = $adapter->copy('file.txt', 'newFile.txt');
        $this->assertTrue($result);
    }
}


src/Adapter.php
<?php

namespace League\Flysystem\WebDAV;

use League\Flysystem\Adapter\AbstractAdapter;
use League\Flysystem\Adapter\Polyfill\NotSupportingVisibilityTrait;
use League\Flysystem\Adapter\Polyfill\StreamedCopyTrait;
use League\Flysystem\Adapter\Polyfill\StreamedTrait;
use League\Flysystem\Config;
use League\Flysystem\Util;
use LogicException;
use Sabre\DAV\Client;
use Sabre\DAV\Exception;

class Adapter extends AbstractAdapter
{
    use StreamedTrait;
    use StreamedCopyTrait;
    use NotSupportingVisibilityTrait;

    /**
     * @var array
     */
    protected static $resultMap = [
        '{DAV:}getcontentlength' => 'size',
        '{DAV:}getcontenttype' => 'mimetype',
        'content-length' => 'size',
        'content-type' => 'mimetype',
    ];

    /**
     * @var Client
     */
    protected $client;

    /**
     * Constructor.
     *
     * @param Client $client
     * @param string $prefix
     */
    public function __construct(Client $client, $prefix = null)
    {
        $this->client = $client;
        $this->setPathPrefix($prefix);
    }

    /**
     * {@inheritdoc}
     */
    public function getMetadata($path)
    {
        $location = $this->applyPathPrefix($path);

        try {
            $result = $this->client->propFind($location, [
                '{DAV:}displayname',
                '{DAV:}getcontentlength',
                '{DAV:}getcontenttype',
                '{DAV:}getlastmodified',
            ]);

            return $this->normalizeObject($result, $path);
        } catch (Exception\FileNotFound $e) {
            return false;
        }
    }

    /**
     * {@inheritdoc}
     */
    public function has($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * {@inheritdoc}
     */
    public function read($path)
    {
        $location = $this->applyPathPrefix($path);

        try {
            $response = $this->client->request('GET', $location);

            if ($response['statusCode'] !== 200) {
                return false;
            }

            return array_merge([
                'contents' => $response['body'],
                'timestamp' => strtotime($response['headers']['last-modified']),
                'path' => $path,
            ], Util::map($response['headers'], static::$resultMap));
        } catch (Exception\FileNotFound $e) {
            return false;
        }
    }

    /**
     * {@inheritdoc}
     */
    public function write($path, $contents, Config $config)
    {
        $location = $this->applyPathPrefix($path);
        $this->client->request('PUT', $location, $contents);

        $result = compact('path', 'contents');

        if ($config->get('visibility')) {
            throw new LogicException(__CLASS__.' does not support visibility settings.');
        }

        return $result;
    }

    /**
     * {@inheritdoc}
     */
    public function update($path, $contents, Config $config)
    {
        return $this->write($path, $contents, $config);
    }

    /**
     * {@inheritdoc}
     */
    public function rename($path, $newpath)
    {
        $location = $this->applyPathPrefix($path);

        try {
            $response = $this->client->request('MOVE', '/'.ltrim($location, '/'), null, [
                'Destination' => '/'.ltrim($newpath, '/'),
            ]);

            if ($response['statusCode'] >= 200 && $response['statusCode'] < 300) {
                return true;
            }
        } catch (Exception\FileNotFound $e) {
            // Would have returned false here, but would be redundant
        }

        return false;
    }

    /**
     * {@inheritdoc}
     */
    public function delete($path)
    {
        $location = $this->applyPathPrefix($path);

        try {
            $this->client->request('DELETE', $location);

            return true;
        } catch (Exception\FileNotFound $e) {
            return false;
        }
    }

    /**
     * {@inheritdoc}
     */
    public function createDir($path, Config $config)
    {
        $location = $this->applyPathPrefix($path);
        $response = $this->client->request('MKCOL', $location);

        if ($response['statusCode'] !== 201) {
            return false;
        }

        return compact('path') + ['type' => 'dir'];
    }

    /**
     * {@inheritdoc}
     */
    public function deleteDir($dirname)
    {
        return $this->delete($dirname);
    }

    /**
     * {@inheritdoc}
     */
    public function listContents($directory = '', $recursive = false)
    {
        $location = $this->applyPathPrefix($directory);

        $response = $this->client->propFind($location, [
            '{DAV:}displayname',
            '{DAV:}getcontentlength',
            '{DAV:}getcontenttype',
            '{DAV:}getlastmodified',
        ], 1);

        array_shift($response);

        $result = [];

        foreach ($response as $path => $object) {
            $path = $this->removePathPrefix($path);
            $object = $this->normalizeObject($object, $path);
            $result[] = $object;

            if ($recursive && $object['type'] === 'dir') {
                $result = array_merge($result, $this->listContents($object['path'], true));
            }
        }

        return $result;
    }

    /**
     * {@inheritdoc}
     */
    public function getSize($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * {@inheritdoc}
     */
    public function getTimestamp($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * {@inheritdoc}
     */
    public function getMimetype($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * Normalise a WebDAV repsonse object.
     *
     * @param array  $object
     * @param string $path
     *
     * @return array
     */
    protected function normalizeObject(array $object, $path)
    {
        if (! isset($object['{DAV:}getcontentlength'])) {
            return ['type' => 'dir', 'path' => trim($path, '/')];
        }

        $result = Util::map($object, static::$resultMap);

        if (isset($object['{DAV:}getlastmodified'])) {
            $result['timestamp'] = strtotime($object['{DAV:}getlastmodified']);
        }

        $result['type'] = 'file';
        $result['path'] = trim($path, '/');

        return $result;
    }
}


src/WebDAVAdapter.php
<?php

namespace League\Flysystem\WebDAV;

use League\Flysystem\Adapter\AbstractAdapter;
use League\Flysystem\Adapter\Polyfill\NotSupportingVisibilityTrait;
use League\Flysystem\Adapter\Polyfill\StreamedCopyTrait;
use League\Flysystem\Adapter\Polyfill\StreamedReadingTrait;
use League\Flysystem\Config;
use League\Flysystem\Util;
use LogicException;
use Sabre\DAV\Client;
use Sabre\DAV\Exception;
use Sabre\DAV\Exception\NotFound;
use Sabre\HTTP\HttpException;

class WebDAVAdapter extends AbstractAdapter
{
    use StreamedReadingTrait;
    use StreamedCopyTrait {
        StreamedCopyTrait::copy as streamedCopy;
    }
    use NotSupportingVisibilityTrait;

    private static $metadataFields = [
        '{DAV:}displayname',
        '{DAV:}getcontentlength',
        '{DAV:}getcontenttype',
        '{DAV:}getlastmodified',
        '{DAV:}iscollection',
    ];

    /**
     * @var array
     */
    protected static $resultMap = [
        '{DAV:}getcontentlength' => 'size',
        '{DAV:}getcontenttype' => 'mimetype',
        'content-length' => 'size',
        'content-type' => 'mimetype',
    ];

    /**
     * @var Client
     */
    protected $client;

    /**
     * @var bool
     */
    protected $useStreamedCopy = true;

    /**
     * Constructor.
     *
     * @param Client $client
     * @param string $prefix
     * @param bool $useStreamedCopy
     */
    public function __construct(Client $client, $prefix = null, $useStreamedCopy = true)
    {
        $this->client = $client;
        $this->setPathPrefix($prefix);
        $this->setUseStreamedCopy($useStreamedCopy);
    }

    /**
     * url encode a path
     *
     * @param string $path
     *
     * @return string
     */
    protected function encodePath($path)
	{
		$a = explode('/', $path);
		for ($i=0; $i<count($a); $i++) {
			$a[$i] = rawurlencode($a[$i]);
		}
		return implode('/', $a);
	}

    /**
     * {@inheritdoc}
     */
    public function getMetadata($path)
    {
        $location = $this->applyPathPrefix($this->encodePath($path));

        try {
            $result = $this->client->propFind($location, self::$metadataFields);

            if (empty($result)) {
                return false;
            }

            return $this->normalizeObject($result, $path);
        } catch (Exception $e) {
            return false;
        } catch (HttpException $e) {
            return false;
        }
    }

    /**
     * {@inheritdoc}
     */
    public function has($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * {@inheritdoc}
     */
    public function read($path)
    {
        $location = $this->applyPathPrefix($this->encodePath($path));

        try {
            $response = $this->client->request('GET', $location);

            if ($response['statusCode'] !== 200) {
                return false;
            }

            return array_merge([
                'contents' => $response['body'],
                'timestamp' => strtotime(is_array($response['headers']['last-modified'])
                    ? current($response['headers']['last-modified'])
                    : $response['headers']['last-modified']),
                'path' => $path,
            ], Util::map($response['headers'], static::$resultMap));
        } catch (Exception $e) {
            return false;
        }
    }

    /**
     * {@inheritdoc}
     */
    public function write($path, $contents, Config $config)
    {
        if (!$this->createDir(Util::dirname($path), $config)) {
            return false;
        }

        $location = $this->applyPathPrefix($this->encodePath($path));
        $response = $this->client->request('PUT', $location, $contents);

        if ($response['statusCode'] >= 400) {
            return false;
        }

        $result = compact('path', 'contents');

        if ($config->get('visibility')) {
            throw new LogicException(__CLASS__.' does not support visibility settings.');
        }

        return $result;
    }

    /**
     * {@inheritdoc}
     */
    public function writeStream($path, $resource, Config $config)
    {
        return $this->write($path, $resource, $config);
    }

    /**
     * {@inheritdoc}
     */
    public function update($path, $contents, Config $config)
    {
        return $this->write($path, $contents, $config);
    }

    /**
     * {@inheritdoc}
     */
    public function updateStream($path, $resource, Config $config)
    {
        return $this->update($path, $resource, $config);
    }

    /**
     * {@inheritdoc}
     */
    public function rename($path, $newpath)
    {
        $location = $this->applyPathPrefix($this->encodePath($path));
        $newLocation = $this->applyPathPrefix($this->encodePath($newpath));

        try {
            $response = $this->client->request('MOVE', '/'.ltrim($location, '/'), null, [
                'Destination' => '/'.ltrim($newLocation, '/'),
            ]);

            if ($response['statusCode'] >= 200 && $response['statusCode'] < 300) {
                return true;
            }
        } catch (NotFound $e) {
            // Would have returned false here, but would be redundant
        }

        return false;
    }

    /**
     * {@inheritdoc}
     */
    public function copy($path, $newpath)
    {
        if ($this->useStreamedCopy === true) {
            return $this->streamedCopy($path, $newpath);
        } else {
            return $this->nativeCopy($path, $newpath);
        }
    }

    /**
     * {@inheritdoc}
     */
    public function delete($path)
    {
        $location = $this->applyPathPrefix($this->encodePath($path));

        try {
            $response =  $this->client->request('DELETE', $location)['statusCode'];


            return $response >= 200 && $response < 300;
        } catch (NotFound $e) {
            return false;
        }
    }

    /**
     * {@inheritdoc}
     */
    public function createDir($path, Config $config)
    {
        $encodedPath = $this->encodePath($path);
        $path = trim($path, '/');

        $result = compact('path') + ['type' => 'dir'];

        if (Util::normalizeDirname($path) === '' || $this->has($path)) {
            return $result;
        }

        $directories = explode('/', $path);
        if (count($directories) > 1) {
            $parentDirectories = array_splice($directories, 0, count($directories) - 1);
            if (!$this->createDir(implode('/', $parentDirectories), $config)) {
                return false;
            }
        }

        $location = $this->applyPathPrefix($encodedPath);
        $response = $this->client->request('MKCOL', $location . $this->pathSeparator);

        if ($response['statusCode'] !== 201) {
            return false;
        }

        return $result;
    }

    /**
     * {@inheritdoc}
     */
    public function deleteDir($dirname)
    {
        return $this->delete($dirname);
    }

    /**
     * {@inheritdoc}
     */
    public function listContents($directory = '', $recursive = false)
    {
        $location = $this->applyPathPrefix($this->encodePath($directory));
        $response = $this->client->propFind($location . '/', self::$metadataFields, 1);

        array_shift($response);
        $result = [];

        foreach ($response as $path => $object) {
            $path = $this->removePathPrefix(rawurldecode($path));
            $object = $this->normalizeObject($object, $path);
            $result[] = $object;

            if ($recursive && $object['type'] === 'dir') {
                $result = array_merge($result, $this->listContents($object['path'], true));
            }
        }

        return $result;
    }

    /**
     * {@inheritdoc}
     */
    public function getSize($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * {@inheritdoc}
     */
    public function getTimestamp($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * {@inheritdoc}
     */
    public function getMimetype($path)
    {
        return $this->getMetadata($path);
    }

    /**
     * @return boolean
     */
    public function getUseStreamedCopy()
    {
        return $this->useStreamedCopy;
    }

    /**
     * @param boolean $useStreamedCopy
     */
    public function setUseStreamedCopy($useStreamedCopy)
    {
        $this->useStreamedCopy = (bool)$useStreamedCopy;
    }

    /**
     * Copy a file through WebDav COPY method.
     *
     * @param string $path
     * @param string $newPath
     *
     * @return bool
     */
    protected function nativeCopy($path, $newPath)
    {
        if (!$this->createDir(Util::dirname($newPath), new Config())) {
            return false;
        }

        $location = $this->applyPathPrefix($this->encodePath($path));
        $newLocation = $this->applyPathPrefix($this->encodePath($newPath));

        try {
            $destination = $this->client->getAbsoluteUrl($newLocation);
            $response = $this->client->request('COPY', '/'.ltrim($location, '/'), null, [
                'Destination' => $destination,
            ]);

            if ($response['statusCode'] >= 200 && $response['statusCode'] < 300) {
                return true;
            }
        } catch (NotFound $e) {
            // Would have returned false here, but would be redundant
        }

        return false;
    }

    /**
     * Normalise a WebDAV repsonse object.
     *
     * @param array  $object
     * @param string $path
     *
     * @return array
     */
    protected function normalizeObject(array $object, $path)
    {
        if ($this->isDirectory($object)) {
            return ['type' => 'dir', 'path' => trim($path, '/')];
        }

        $result = Util::map($object, static::$resultMap);

        if (isset($object['{DAV:}getlastmodified'])) {
            $result['timestamp'] = strtotime($object['{DAV:}getlastmodified']);
        }

        $result['type'] = 'file';
        $result['path'] = trim($path, '/');

        return $result;
    }

    private function isDirectory(array $object)
    {
        return isset($object['{DAV:}iscollection']) && $object['{DAV:}iscollection'] === '1';
    }
}
